src/bc_protocol/media_packet.rs

```diff
@@ -0,0 +1,381 @@
+            2 => u16::from_le_bytes(bytes.try_into().expect("slice with incorrect length"))
+                .try_into()
+                .expect("u16 won't fit into usize"),
+            1 => u8::from_le_bytes(bytes.try_into().expect("slice with incorrect length"))
+                .try_into()
+                .expect("u8 won't fit into usize"),
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn kind_from_raw(data: &[u8]) -> MediaDataKind {
+        // When calling this ensure you have enough data for header_size +2
+        // Else full_header_check_from_kind will fail because we check the
+        // First two bytes after the header for the audio stream
+        // Since AAC and ADMPC streams start in a predicatble manner
+        if data.len() < 4 {
+            return MediaDataKind::Invalid;
+        }
+        const MAGIC_VIDEO_INFO_V1: &[u8] = &[0x31, 0x30, 0x30, 0x31];
+        const MAGIC_VIDEO_INFO_V2: &[u8] = &[0x31, 0x30, 0x30, 0x32];
+        const MAGIC_AAC: &[u8] = &[0x30, 0x35, 0x77, 0x62];
+        const MAGIC_ADPCM: &[u8] = &[0x30, 0x31, 0x77, 0x62];
+        const MAGIC_IFRAME: &[u8] = &[0x30, 0x30, 0x64, 0x63];
+        const MAGIC_PFRAME: &[u8] = &[0x30, 0x31, 0x64, 0x63];
+
+        let magic = &data[..4];
+        let kind = match magic {
+            MAGIC_VIDEO_INFO_V1 | MAGIC_VIDEO_INFO_V2 => MediaDataKind::InfoData,
+            MAGIC_AAC => MediaDataKind::AudioDataAac,
+            MAGIC_ADPCM => MediaDataKind::AudioDataAdpcm,
+            MAGIC_IFRAME => MediaDataKind::VideoDataIframe,
+            MAGIC_PFRAME => MediaDataKind::VideoDataPframe,
+            _ if data.len() == CHUNK_SIZE => MediaDataKind::Continue,
+            _ => {
+                trace!("Unknown magic kind: {:x?}", &magic);
+                MediaDataKind::Unknown
+            }
+        };
+
+        // I've never had this fail yet. It checks more bytes then just the magic
+        // Including some 2 bytes at the start of the data
+        if !MediaData::full_header_check_from_kind(kind, &data) {
+            return MediaDataKind::Invalid;
+        } else {
+            return kind;
+        }
+    }
+
+    pub fn kind(&self) -> MediaDataKind {
+        MediaData::kind_from_raw(&self.data)
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+
+    pub fn full_header_check_from_kind(kind: MediaDataKind, data: &[u8]) -> bool {
+        // This will run more advanced checks to ensure it is a valid header
+        let header_size = MediaData::header_size_from_kind(kind);
+        if data.len() < header_size {
+            trace!("Magic failed header checks on size: {:x?}", &data);
+            return false;
+        }
+
+        match kind {
+            MediaDataKind::VideoDataIframe => {
+                let stream_type = &data[4..8];
+                if let Ok(stream_type_name) = str::from_utf8(stream_type) {
```

Comment:
It shouldn't be necessary to convert this to UTF8. You can simply match against the bytes, `b"H264"`, which is just sugar over the byte array.

Reply:
Done